example/build_cmake_installed_different_compiler/CMakeLists.txt

```diff
@@ -6,16 +6,20 @@ cmake_minimum_required(VERSION 3.16)
 # Kokkos flags will only apply to C++ files
 project(Example CXX Fortran)
 
+find_package(Kokkos REQUIRED)
 # Look for an installed Kokkos but force using the compiler launcher
 # to ensure that targets depending on Kokkos use the same compiler
-# as when kokkos was installed, e.g. if kokkos was built with
-# g++ and the CMAKE_CXX_COMPILER=clang++ then example_with_kokkos
-# will be compiled and linked with g++ whereas example_no_kokkos
-# will be compiled and linked with clang++
-find_package(Kokkos REQUIRED COMPONENTS launch_compiler)
+# as when kokkos was installed.
+set(Kokkos_FORCE_COMPILER_LAUNCHER true)
+# Alternatively, this can be input at configure time with:
+# -DKokkos_FORCE_COMPILER_LAUNCHER=ON
```

Comment:
Should we be recommending to set this variable before finding Kokkos?
I know in practice it does not matter at the moment but, conceptually, isn't it something like a hint from the user saying what compatible Kokkos to find?

Reply:
Yeah, the value actually has to be available before the call to `find_package`. It is setting how Kokkos should behave in the current build. We could force the user to select this at configure time (of Kokkos itself) to avoid this awkwardness. But this might double the configs that need to be installed.

If we use the variable at configure time (of the lib that consumes Kokkos) like here, we avoid the duplicate installs but require the user to adhere to providing it pre calling `find_package`.

This goes partly away with https://github.com/kokkos/kokkos/pull/7582 that restricts `find_package(Kokkos)` to only set the compiler/language for files that are actually linking to Kokkos. And if we use the CMake language feature as default it goes away completely (and we could think about removing the compiler launcher).

Reply:
>  We could force the user to select this at configure time (of Kokkos itself) to avoid this awkwardness. But this might double the configs that need to be installed.

For people using a package manager to install software, however, this makes more sense.

> If we use the variable at configure time (of the lib that consumes Kokkos) like here, we avoid the duplicate installs but require the user to adhere to providing it pre calling `find_package`.

Which is not so easy to ensure for the software stack. Does this mean it should be set in `FooConfig.cmake` when doing `find_dependency(Kokkos)?`

Reply:
> For people using a package manager to install software, however, this makes more sense.

hmm but this would just result in 4x the configurations in the package manager. Not sure if we really want this.
And packages would need to specify which config of Kokkos they are fine with.

> Which is not so easy to ensure for the software stack. Does this mean it should be set in `FooConfig.cmake` when doing `find_dependency(Kokkos)?`

yes, it would be required. And yes you are right, this is a problem (also for the previous solution using the components). As `find_dependency` optimizes out repeated `find_package` calls it is not guaranteed that the behavior users specify (with the variables introduced in this PR) is actually executed. This can definitely break things. 
However, this is already a problem with how we do things now. e.g.
```
find_package(Kokkos COMPONENTS separable_compilation)
find_package(KokkosKernels) #internally does 'find_package(Kokkos)' thus invalidating the `separable_compilation`
```

In the long run I think we should move away from how we do things now (setting anything globally or forcing some compiler to be used) and tell users to spell out on which targets they want a specific Kokkos to set the compiler.
https://github.com/kokkos/kokkos/pull/7582 does a step in this direction.